cmd/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	if newVersion == "" {
+		var err error
+		_, newVersion, _, err = versionCmd.GetVersion(siteURL + "/version.txt")
+		if err != nil {
+			return errors.Wrap(err, "unable to detect new version")
+		}
+	}
+
+	if newVersion == fs.Version {
+		fmt.Println("rclone is up to date")
+		return nil
+	}
+
+	// Find the executable path
+	executable, err := os.Executable()
+	if err != nil {
+		return errors.Wrap(err, "unable to find executable")
+	}
+
+	targetFile := opt.Output
+	if targetFile == "" {
+		targetFile = executable
+	}
+
+	// Check for possible access errors in advance
+	newFile := targetFile + ".new"
+	savedFile := ""
+	if runtime.GOOS == "windows" {
+		savedFile = targetFile + ".old"
+	}
+
+	if savedFile == executable || newFile == executable {
+		return fmt.Errorf("%s: a temporary file would overwrite executable, specify a different --output", targetFile)
+	}
+
+	if err := verifyAccess(targetFile); err != nil {
+		return err
+	}
+
+	// Download the update as a temporary file
+	if err := downloadUpdate(opt.Beta, newVersion, siteURL, newFile); err != nil {
+		return errors.Wrap(err, "failed to update rclone")
+	}
+
+	// Copy permission bits from the old executable
+	fileMode := os.FileMode(0755)
+	if fileInfo, err := os.Lstat(targetFile); err == nil {
+		fileMode = fileInfo.Mode()
+	}
+	if err := os.Chmod(newFile, fileMode); err != nil {
+		return errors.Wrap(err, "failed to set permissions")
+	}
+
+	// Replace current executable by the new file
+	if err = os.Remove(targetFile); os.IsNotExist(err) {
+		err = nil
+	}
+	if err != nil && savedFile != "" {
+		// Windows forbids removal of a running executable so we rename it
+		var saveErr error
+		if saveErr = os.Remove(savedFile); os.IsNotExist(saveErr) {
+			saveErr = nil
+		}
+		if saveErr == nil {
+			saveErr = os.Rename(targetFile, savedFile)
+		}
+		if saveErr == nil {
+			fmt.Printf("The old executable was saved as %s\n", savedFile)
+			err = nil
+		} else {
+			// The rename trick didn't work out, proceed like on Unix
+			_ = os.Remove(savedFile)
+		}
+	}
+	if err == nil {
+		err = os.Rename(newFile, targetFile)
+	}
+	if err != nil {
+		_ = os.Remove(newFile)
+		return err
+	}
+	fmt.Printf("Successfully updated rclone to version %s\n", newVersion)
+	return nil
+}
+
+func downloadUpdate(beta bool, version, siteURL, newFile string) error {
+	archiveFilename := fmt.Sprintf("rclone-%s-%s-%s.zip", version, runtime.GOOS, runtime.GOARCH)
+	archiveURL := fmt.Sprintf("%s/%s/%s", siteURL, version, archiveFilename)
+	archiveBuf, err := downloadFile(archiveURL)
+	if err != nil {
+		return err
+	}
+	fs.Debugf(nil, "downloaded release archive: %s", archiveURL)
+	gotHash := sha256.Sum256(archiveBuf)
+	fs.Debugf(nil, "archive hashsum: %s", hex.EncodeToString(gotHash[:]))
+
+	// CI/CD does not provide hashsums for beta releases
+	if !beta {
+		hashsumsURL := fmt.Sprintf("%s/%s/SHA256SUMS", siteURL, version)
+		hashsumsBuf, err := downloadFile(hashsumsURL)
```

Comment:
In a perfect world we'd check the signature on SHA256SUMS - it is gpg signed with my pgp key. We'd need to embed the PGP signature in rclone I think to make this work securely.

Let's leave this out for the moment.

Reply:
Embedding gpg key will require tweaks in CI/CD (github secret -> workflow variable -> golang linker key -> executable)
I will submit it as a separate PR.

Reply:
It would only need the public key embedded, not the private key so that could just go in the code as a text blob.

However this should really be done with an `rclone release key` rather than my private key.

There is probably a much better way of doing it other than using GPG too - something much lighter weight.

Reply:
I don't know what your public key is
I tried this one https://www.craig-wood.com/nick/pub/pgp-key.txt
But golang cannot read it. I think PGP-2 isn't supported.
We'll need a modern public key.

```
package main

import "fmt"
import "os"
import "golang.org/x/crypto/openpgp"

func main() {
        in, err := os.Open("pgp-key.txt")
        _, err = openpgp.ReadArmoredKeyRing(in)
        fmt.Println(err)
}
```

Reply:
@ncw
we could make this world perfect 🐇 🌈 by finding such a public release key that passes the test above ✨